File: tools/github/require-needs-success.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Jobs in the `test-report` dependency list that skip on every pull request by
 * design, mapped to the reason they skip.
 *
 * `test-report` itself only runs on `pull_request`, so this is the complete set
 * of legitimate skips: both entries are gated by
 * `if: ${{ github.event_name != 'pull_request' }}` in `.github/workflows/check.yml`
 * and therefore never run in the event that `test-report` aggregates.
 *
 * Every other needed job runs on pull requests, so a `skipped` result there means
 * the job never ran at all — `playground-test`, for instance, is skipped when the
 * `build-js-packages` job it needs fails. Those skips are failures for gating
 * purposes: the work they were meant to gate was not done.
 *
 * `tests/tooling/github-workflows-check-gate.test.ts` re-derives this list from
 * the workflow's `if:` guards, so adding a pull-request-skipping job to
 * `test-report`'s `needs:` fails that test until the job is classified here.
 */
export const PULL_REQUEST_SKIPPED_JOBS = Object.freeze({
  "nix-flake": "runs on push and schedule only",
  "source-coverage": "runs on push and schedule only",
});

function sortedEntries(needs) {
  return Object.entries(needs).sort(([left], [right]) =>
    left < right ? -1 : left > right ? 1 : 0,
  );
}

function formatNames(names) {
  return names.join(", ");
}

/**
 * Decides whether the `test-report` aggregator may report success.
 *
 * `test-report` runs with `if: always()` so that it still collects the test
 * inventory when a dependency fails. Without this check its own steps succeed
 * and the required status check reports success while any of the jobs it
 * aggregates is red, which makes the whole check suite advisory.
 *
 * @param {Record<string, { result?: string }>} needs Parsed `toJSON(needs)` payload.
 * @param {Record<string, string>} skippableJobs Jobs allowed to report `skipped`.
 * @returns {{ exitCode: number, message: string }}
 */
export function aggregateNeedsResults(needs, skippableJobs = PULL_REQUEST_SKIPPED_JOBS) {
  if (needs === null || typeof needs !== "object" || Array.isArray(needs)) {
    throw new Error("The needs context must be an object of job results");
  }
  const entries = sortedEntries(needs);
  if (entries.length === 0) {
    throw new Error("The needs context is empty: test-report must depend on the jobs it gates");
  }

  const succeeded = [];
  const skippedByDesign = [];
  const unresolved = [];
  for (const [job, value] of entries) {
    const result = value === null || typeof value !== "object" ? undefined : value.result;
    if (typeof result !== "string" || result === "") {
      throw new Error(`Job ${job} reported no result in the needs context`);
    }
    if (result === "success") {
      succeeded.push(job);
    } else if (result === "skipped" && Object.hasOwn(skippableJobs, job)) {
      skippedByDesign.push(job);
    } else {
      unresolved.push(`  - ${job}: ${result}`);
    }
  }

  const allowed = formatNames(Object.keys(skippableJobs).sort());
  if (unresolved.length > 0) {
    return {
      exitCode: 1,
      message: [
        `test-report gate: ${unresolved.length} of ${entries.length} needed jobs did not succeed.`,
        ...unresolved,
        "test-report is a required status check, so it must not pass while a job it aggregates is red.",
        `Only these jobs may skip on a pull request: ${allowed}.`,
      ].join("\n"),
    };
  }

  const skipped =
    skippedByDesign.length === 0
      ? "0 skipped on a pull request by design."
      : `${skippedByDesign.length} skipped on a pull request by design: ${formatNames(skippedByDesign)}.`;
  return {
    exitCode: 0,
    message: `test-report gate: all ${entries.length} needed jobs are accounted for. ${succeeded.length} succeeded; ${skipped}`,
  };
}

function main() {
  const raw = process.env.NEEDS_JSON;
  if (!raw) {
    console.error(
      "NEEDS_JSON is required: pass ${{ toJSON(needs) }} to tools/commands/ci/github/require-needs-success.rs",
    );
    process.exitCode = 1;
    return;
  }
  const { exitCode, message } = aggregateNeedsResults(JSON.parse(raw));
  if (exitCode === 0) {
    console.log(message);
    return;
  }
  console.error(message);
  process.exitCode = exitCode;
}

const entrypoint = process.argv[1]
  ? fileURLToPath(import.meta.url) === path.resolve(process.argv[1])
  : false;
if (entrypoint) {
  main();
}

File: tools/commands/ci/github/require-needs-success.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! serde_json = "1"
//!
//! [package]
//! edition = "2024"
//! ```

use serde_json::Value;
use std::{collections::BTreeMap, env, process::ExitCode};

const SKIPPABLE_JOBS: &[(&str, &str)] = &[
    ("nix-flake", "runs on push and schedule only"),
    ("source-coverage", "runs on push and schedule only"),
];

fn main() -> ExitCode {
    let Some(raw) = env::var_os("NEEDS_JSON") else {
        eprintln!(
            "NEEDS_JSON is required: pass ${{{{ toJSON(needs) }}}} to tools/commands/ci/github/require-needs-success.rs"
        );
        return ExitCode::from(1);
    };
    let raw = raw.to_string_lossy();
    match aggregate(&raw) {
        Ok((0, message)) => {
            println!("{message}");
            ExitCode::SUCCESS
        }
        Ok((_, message)) => {
            eprintln!("{message}");
            ExitCode::from(1)
        }
        Err(error) => {
            eprintln!("{error}");
            ExitCode::from(1)
        }
    }
}

fn aggregate(raw: &str) -> Result<(u8, String), String> {
    let json: Value = serde_json::from_str(raw).map_err(|error| error.to_string())?;
    let needs = json
        .as_object()
        .ok_or_else(|| "The needs context must be an object of job results".to_string())?;
    if needs.is_empty() {
        return Err(
            "The needs context is empty: test-report must depend on the jobs it gates".to_string(),
        );
    }

    let skippable = SKIPPABLE_JOBS.iter().copied().collect::<BTreeMap<_, _>>();
    let mut succeeded = Vec::new();
    let mut skipped_by_design = Vec::new();
    let mut unresolved = Vec::new();

    for (job, value) in needs.iter().collect::<BTreeMap<_, _>>() {
        let result = value
            .get("result")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty())
            .ok_or_else(|| format!("Job {job} reported no result in the needs context"))?;
        if result == "success" {
            succeeded.push(job.as_str());
        } else if result == "skipped" && skippable.contains_key(job.as_str()) {
            skipped_by_design.push(job.as_str());
        } else {
            unresolved.push(format!("  - {job}: {result}"));
        }
    }

    let allowed = skippable.keys().copied().collect::<Vec<_>>().join(", ");
    if !unresolved.is_empty() {
        let mut lines = vec![format!(
            "test-report gate: {} of {} needed jobs did not succeed.",
            unresolved.len(),
            needs.len()
        )];
        lines.extend(unresolved);
        lines.push(
            "test-report is a required status check, so it must not pass while a job it aggregates is red."
                .to_string(),
        );
        lines.push(format!(
            "Only these jobs may skip on a pull request: {allowed}."
        ));
        return Ok((1, lines.join("\n")));
    }

    let skipped = if skipped_by_design.is_empty() {
        "0 skipped on a pull request by design.".to_string()
    } else {
        format!(
            "{} skipped on a pull request by design: {}.",
            skipped_by_design.len(),
            skipped_by_design.join(", ")
        )
    };
    Ok((
        0,
        format!(
            "test-report gate: all {} needed jobs are accounted for. {} succeeded; {skipped}",
            needs.len(),
            succeeded.len()
        ),
    ))
}
